src/controller/metakg.py

```diff
@@ -0,0 +1,34 @@
+import logging
+from collections.abc import Mapping
+
+from model import MetaKGDoc
+from utils.metakg.parser import Parser
+from .base import AbstractWebEntity, decoder
+
+logger = logging.getLogger(__name__)
+
+
+class MetaKGEntity(AbstractWebEntity, Mapping):
+    LOOKUP_FIELDS = []
+    MODEL_CLASS = MetaKGDoc
+
+    @classmethod
+    def create_by_smartapis(cls, smartapi_entities, include_reasoner=False):
+        for smartapi_entity in smartapi_entities:
```

Comment:
since we are now in the SmartAPI application itself. I don't see any reason we want to process metadata in a batch here as `smartapi_entities`. We can just loop through docs from SmartAPI model, and handle one API at a time:

```
for smartapi_entity in smartapi_all:
     if tags.name contains translator:
          if TRAPI API:
                parser.get_TRAPI_metadatas
          elif x_bte API:
                parser.get_non_TRAPI_metadatas
          else:
                just skip
          bulk_save all metadata from one API
```

Reply:
for easier debugging and cleaner code structure, you might want to have a method of parsing out metakg metadata for a given API id, which covers the above `if...elif...else` block:

```
parser.get_metakg(smartapi_id)
```

Reply:
> since we are now in the SmartAPI application itself. I don't see any reason we want to process metadata in a batch here as smartapi_entities. We can just loop through docs from SmartAPI model, and handle one API at a time:

Of course, I loaded and process metadata for all stored SmartAPI. but I just think doing in batch to prevent loading too many data into memory at a time, so I split them into small set, and process each one in a batch.

If you don't want to do that, I will remove the batch logic, and process all SmartAPI in one time.

@newgene

Reply:
@sengineer0 I actually think the opposite. Because we are within the SmartAPI application, so you can just process one SmartAPI entry at a time, no need to be in a batch. That way we can further reduce memory usage. Definitely not loading all entries in memory and processing them in one batch, which will make it worst :-)

However, I think for the controller method, you can keep ```create_by_smartapis``` but create another one ```create_by_smartapi``` for processing one at a time, even another method like `create_by_smartapi_id`. These are very useful for debugging purposes.

Reply:
that makes seen. I will update as you recommended :)

@newgene

Reply:
it's updated @newgene